Implement multiple dispatch in boxed c10 dispatcher

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26719 Tests for fallback boxed dispatch.
* #26368 Add boxed fallback
* #26367 Refactor dispatch structure so fallback code lives inline.
* #26361 Change calling convention of ATenDispatch from getOp to callUnboxed.
* #26718 Add some missing constructors to IValue.
* #26360 Remove unnecessary include from TensorBody
* **#26118 Implement multiple dispatch in boxed c10 dispatcher**
* #25914 Delete backwards compatibility Backend overload for registerOp

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

ezyang committed:
Implement multiple dispatch in boxed c10 dispatcher

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -205,40 +203,36 @@ class DispatchTable final {
+      TensorTypeSet ts;
+      for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+        if (ivalue.isTensor()) {
+          ts = ts | ivalue.toTensor().type_set();
```
This does a refcount bump. In my dispatch benchmarks, that was deadly for perf, which is why I the original code used `unsafeToTensorImpl()` to avoid the refcount bump. You should probably do that here too.

ezyang replied:
I originally wrote `toTensorRef` here and was shocked that this didn't exist

smessmer replied:
`toTensorRef` can't exist because `IValue` stores `TensorImpl` and not `Tensor`, so if you want a `const Tensor&`, you have to create a `Tensor` object first. You can only get `toTensorImplRef`, which is what `unsafeToTensorImpl` is doing.

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -205,40 +203,36 @@ class DispatchTable final {
   };
 
   static DispatchStrategy get_dispatch_strategy_(const FunctionSchema& schema) {
+    bool is_valid = false;
```
nit: This code can become shorter by removing `is_valid` and just returning from the loop. Up to you though, the current implementation is also ok.

smessmer reviewed (approved):
Looks good :)

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -205,40 +203,36 @@ class DispatchTable final {
+
+      TensorTypeSet ts;
+      for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+        if (ivalue.isTensor()) {
```
A `C10_LIKELY` might help with perf a bit here since tensors are more likely than tensor lists.

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

ezyang committed:
Update on "Implement multiple dispatch in boxed c10 dispatcher"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17404367](https://our.internmc.facebook.com/intern/diff/D17404367)

[ghstack-poisoned]

CaoZhongZ commented:
Quick question, what does boxed or unboxed mean?

ezyang commented:
@CaoZhongZ boxed means we have a uniform runtime representation and can write code that generically operates over it. E.g., this is how reflection in Java works. In C++, all representations are unboxed, which means you can have more efficient representations, but it's a lot harder to write generic code without codegen (templates, e.g.)

CaoZhongZ commented:
@ezyang I got it, after digest a while. Thanks!
